beacon-chain/slasher/detect_attestations.go

```diff
@@ -5,74 +5,175 @@ import (
+func (s *Service) detectSlashableAttestations(
+	ctx context.Context,
+	opts *chunkUpdateOptions,
+	attestations []*slashertypes.CompactAttestation,
+) error {
+	_, err := s.updateSpans(ctx, &chunkUpdateOptions{
+		kind:                slashertypes.MinSpan,
+		validatorChunkIndex: opts.validatorChunkIndex,
+		currentEpoch:        opts.currentEpoch,
+	})
+	if err != nil {
+		return err
+	}
+	_, err = s.updateSpans(ctx, &chunkUpdateOptions{
+		kind:                slashertypes.MaxSpan,
+		validatorChunkIndex: opts.validatorChunkIndex,
+		currentEpoch:        opts.currentEpoch,
+	})
+	if err != nil {
+		return err
+	}
 
+	// Update the latest written epoch for all involved validator indices.
+	validatorIndices := s.params.validatorIndicesInChunk(opts.validatorChunkIndex)
+	return s.serviceCfg.Database.SaveLatestEpochAttestedForValidators(ctx, validatorIndices, opts.currentEpoch)
 }
 
-// Group a list of attestations into batches by validator chunk index.
-// This way, we can detect on the batch of attestations for each validator chunk index
-// concurrently, and also allowing us to effectively use a single 2D chunk
-// for slashing detection through this logical grouping.
-func (s *Service) groupByValidatorChunkIndex(
-	attestations []*slashertypes.CompactAttestation,
-) map[uint64][]*slashertypes.CompactAttestation {
-	groupedAttestations := make(map[uint64][]*slashertypes.CompactAttestation)
-	for _, att := range attestations {
-		validatorChunkIndices := make(map[uint64]bool)
-		for _, validatorIdx := range att.AttestingIndices {
-			validatorChunkIndex := s.params.validatorChunkIndex(types.ValidatorIndex(validatorIdx))
-			validatorChunkIndices[validatorChunkIndex] = true
+// Updates spans and detects any slashable attester offenses along the way.
+// 1. Update the latest written epoch for all validator indices involved up and
+//    including the current epoch, return the updated chunks by chunk index.
+// 2. Using the chunks from step (1):
+//      for every attestation by chunk index:
+//        for each validator in the attestation's attesting indices:
+//          - Check if the attestation is slashable, if so return a slashing object.
+// 3. Save the updated chunks to disk.
+func (s *Service) updateSpans(
+	ctx context.Context,
+	opts *chunkUpdateOptions,
+) (slashertypes.SlashingKind, error) {
+	// Determine the chunk indices we need to use for slashing detection.
+	validatorIndices := s.params.validatorIndicesInChunk(opts.validatorChunkIndex)
+	chunkIndices, err := s.determineChunksToUpdateForValidators(ctx, opts, validatorIndices)
+	if err != nil {
+		return slashertypes.NotSlashable, err
+	}
+	// Load the required chunks from disk.
+	chunksByChunkIdx, err := s.loadChunks(ctx, opts, chunkIndices)
+	if err != nil {
+		return slashertypes.NotSlashable, err
+	}
+
+	// Apply the attestations to the related chunks.
+	// TODO(#8331): Apply...
```

Comment:
The actual logic of applying an attestation to the min and max spans will be done in a separate PR